### ISSUE TITLE
Always return false positive with ternary

### ISSUE THREAD

**Issue by ChristianMurphy** [xjamundx/eslint-plugin-promise#31]
ESLint Version: `3.8.1`
Promise Plugin Version: `3.0.0`

Configuration

``` json
{
  "plugins": ["promise"],
  "rules": {
    "promise/always-return": "error"
  }
}
```

Minimal example

``` js
something.then(function (a) {
  return a ? 'answer' : 42;
});
```

Expected: No error
Actual: `error  Each then() should return a value or throw`

**Comment by xjamundx** [comment]
Great suggestion, can you work on the pull request?

**Comment by r-murphy** [comment]
I have a potential fix for this on my fork, but I'm worried about unforeseen side effects due to my limited understanding of AST. 

I solved it by propagating the 'good' property down the segment branches from parent to children.
So once `return` or `throw` is found and marked as `good`, all the child segments of that one get marked as good also.

The reason I took this approach is to handle ternary expressions nested inside of an object literal which were getting marked as errors.

For example:
```js
return {
  foo: {
    bar: (x ? 1 : 0)
  }
}
```

I considered checking if the last line of the function was a return or throw (old logic) and then falling back to the new branch based logic but that wouldn't work in this more complex example that has branching and ternary mixed:

```js
if (useBool) {
  return x ? 1 : 0; // Simple ternary.
}
else {
  return {
    bar: (x ? 1 : 0) // Ternary in object literal.
  }
}
```

I *think* it's safe since the `prevSegments` of a function is empty. So the `good` doesn't propagate from outer function to inner function if someone happens to structure their promise chain that way.

**Comment by geoffdutton** [comment]
I've submitted a PR for a potential fix, any thoughts? #56

**Comment by xjamundx** [comment]
Thanks for submitting. Let me take a look this weekend

**Comment by ChristianMurphy** [comment]
resolved by #56
